queue/byte_queue.go

```diff
@@ -0,0 +1,144 @@
+package queue
+
+import (
+	"sync"
+
+	"github.com/garyburd/redigo/redis"
+)
+
+// ByteQueue represents either a FILO or FIFO queue contained in a particular
+// Redis keyspace. It allows callers to push `[]byte` payloads, and receive them
+// back over the `In() <-chan []byte`. It is typically used in a distributed
+// setting, where the pusher may not always get the item back.
+type ByteQueue struct {
+	in     chan []byte
+	closer chan struct{}
+	errs   chan error
+	wg     sync.WaitGroup
+
+	name string
+	pool *redis.Pool
+
+	pmu       sync.RWMutex
+	processor Processor
+}
+
+// NewByteQueue allocates and returns a pointer to a new instance of a
+// ByteQueue. It initializes itself using the given *redis.Pool, and the name,
+// which refers to the keyspace wherein these values will be stored.
+//
+// Internal channels are also initialized here.
+func NewByteQueue(pool *redis.Pool, name string) *ByteQueue {
+	return &ByteQueue{
+		in:     make(chan []byte),
+		closer: make(chan struct{}, 2),
+		errs:   make(chan error),
+
+		name: name,
+		pool: pool,
+	}
+}
+
+// Start spawns a goroutine that will populate the `In() <-chan []byte`. If
+// Start is not called, items still may be `Push`-ed, but they will not show up
+// in the In() channel, and attempting to <-In() will cause a panic.
+//
+// The spawned goroutine is automatically halted when `Close()` is called.
+func (b *ByteQueue) Start() {
+	go b.untilClosed(b.pull)
+}
```

Comment:
Maybe call this `StartRecv` or something like that? If I just want to _send_ stuff to a queue, it's not clear at first glance that I shouldn't Start the queue.